### PR TITLE
Update development process to use Conventional Commits

### PR DESCRIPTION
**Pull request by kgryte**
## Description

> What is the purpose of this pull request?

This pull request:

- updates Git commit guidance to use [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/), which will allow us to eventually automate release and changelog generation for both the main project and all standalone repositories. This migration presents a sizable change from the current practice of relatively freeform commits to a practice in which commits must adhere to a fairly strict structure.
- updates contribution and development guidance. Notably, the development guide now includes a warning regarding limiting the clone depth, in light of recent difficulties encountered by new contributors wanting to rebase on the latest `develop` branch.
- adds license headers to various contribution guides.
- updates copy based on current prose conventions (e.g., instead of "beep boop; e.g., foo bar", do "beep boop (e.g., foo bar)").
- includes a link to the doctest guide in the development guide.
- disables ESLint rules where examples are intended to be illustrative.
- adds guidance for authoring `fixup!` commits.

## Related Issues

> Does this pull request have any related issues?

None.

## Questions

> Any questions for reviewers of this pull request?

- Are we satisfied with the list of commit "types" as set forth in the Git commit guide? Too many? Too few?
- Should we update our contributing guide in other ways?

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

- Provided this PR is accepted, we'll add support for commit message linting to enforce adherence to the new conventions.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

cc @Planeshifter @steff456 @Pranavchiku 

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

File: docs/development.md [hunk 1]
```diff
@@ -99,15 +109,15 @@ If you are wanting to contribute to stdlib, first [fork][github-fork] the reposi
...
+where `<username>` is your GitHub username (assuming you are using GitHub to manage public repositories). The repository has a large commit history, leading to slow download times. You can reduce the download time by limiting the clone [depth][git-clone-depth].
 
 <!-- run-disable -->
 
 ```bash
 $ git clone --depth=<depth> https://github.com/<username>/stdlib.git
 ```
 
-where `<depth>` refers to the number of commits you want to download (as few as `1` and as many as the entire project history).
+where `<depth>` refers to the number of commits you want to download (as few as `1` and as many as the entire project history). **However, you should be aware that limiting clone depth can cause difficulties later when attempting to rebase a pull request on the latest development branch.** For simple pull requests, limiting clone depth is likely to work out fine; however, for more complex pull requests, including those depending on upstream changes, limiting clone depth may be a source of Git errors (e.g., due to unrelated Git histories), and, thus, you may be forced to re-clone the repository and start over.
```
Comment (Pranavchiku) [inline docs/development.md]:
```suggestion
where `<depth>` refers to the number of commits you want to download (as few as `1` and as many as the entire project history). **However, limiting clone depth can cause difficulties later when attempting to rebase a pull request on the latest development branch.** For simple pull requests, limiting clone depth is likely to work out fine; however, for more complex pull requests, including those depending on upstream changes, limiting clone depth may be a source of Git errors (e.g., due to unrelated Git histories), and, thus, you may be forced to re-clone the repository and start over.
```

File: docs/style-guides/git/README.md [hunk 2]
```diff
@@ -1,199 +1,289 @@
...
+-   `Reviewed-by:` the name and e-mail of a commit reviewer (e.g., `Jane Doe <jane@doe.com>`).
+-   `Co-authored-by:` the name and e-mail of a contributor who collaborated on the commit.
+-   `Ref:` link to a public resource (e.g., commit SHA, external documentation, reference implementation, etc).
+-   `Private-ref:` link to a private resource (e.g., issue tracker, pull request, discussion, etc) related to the commit. Private references should only be included in the commit message body using the `Private-ref` token.
+
+Each token may be repeated; however, only one token is allowed per line. Tokens **must** begin with a capital letter followed by all lowercase letters, a colon `:`, and a space. A token **must** not include whitespace, and individual words **must** be separated by a hyphen `-` (e.g., `Reviewed-by`, not `Reviewed by`).
+
+The only exceptions to the restrictions on token case and inclusion of whitespace are `BREAKING CHANGE`, which must be in uppercase and include a single whitespace character, and `DEPRECATED`, which must be in uppercase. Both `BREAKING CHANGE` and `DEPRECATED` **must** come before all other tokens.
```
Comment (Pranavchiku) [inline docs/style-guides/git/README.md]:
```suggestion
The only exceptions to the restrictions on token case and inclusion of whitespace are `BREAKING CHANGE`, which **must** be in uppercase and include a single whitespace character, and `DEPRECATED`, which **must** be in uppercase. Both `BREAKING CHANGE` and `DEPRECATED` **must** come before all other tokens.
```

**Review by Pranavchiku — APPROVED** [review]
Looks good to me!

**Comment by kgryte** [comment]
Thanks for reviewing, @Pranavchiku!

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -136,43 +156,51 @@ $ git remote add upstream git://github.com/stdlib-js/stdlib.git
...
 
+<!-- run-disable -->
+
 ```bash
 $ git add files/which/changed
 $ git commit
 ```
 
-When writing commit messages, follow the Git [style guide][stdlib-style-guides].
+When writing commit messages, follow the Git [style guide][stdlib-style-guides-git]. Adherence to project commit conventions is necessary for project automation which automatically generates release notes and changelogs from commit messages.
```
Comment (Planeshifter) [inline CONTRIBUTING.md:191]: Should add a comma before "which".

**Review by Planeshifter — APPROVED** [review]
Here we come, Conventional Commits!

File: docs/style-guides/git/README.md [hunk 4]
```diff
@@ -1,199 +1,289 @@
...
+<header>
+
+<body>
+
+<footer>
+```
+
+The **header** is **mandatory** and must conform to the [commit message header](#commit-message-header) format.
+
+The **body** is **mandatory** for all non-trivial commits. When present, the body must be at least 20 characters in length and must conform to the [commit message body](#commit-message-body) format.
```
Comment (steff456) [inline docs/style-guides/git/README.md:53]: I think maybe including an example of a non-trivial commit may clarify what's the severity of the fix and the case that we should be looking to add a body to our commits from now on.

**Review by steff456 — CHANGES_REQUESTED** [review]
I just left a minor comment, otherwise it looks good to me!

We are going to need a ramp up time to get used to the new commits. Do y'all think we should give feedback after a while to see how we feel with all the options?

File: docs/style-guides/git/README.md (hunk 4 above)
Comment (kgryte) [reply docs/style-guides/git/README.md:53]: Agreed. My plan is to add a list of user stories showing commit examples. Will try to get to later this evening/weekend.

**Comment by kgryte** [comment]
@steff456 Indeed. The plan is to trial this for a while to collect feedback, etc, before going "live" with tooling and automation.

**Comment by steff456** [comment]
Do you have a preference on where the feedback should go?

**Comment by kgryte** [comment]
@steff456 Maybe we can just use this PR? So if you have further thoughts, etc, just leave a comment here or on the relevant line of this PR?

**Comment by steff456** [comment]
Excellent! Sounds like a plan 😬

File: CONTRIBUTING.md (hunk 3 above)
Comment (kgryte) [reply CONTRIBUTING.md:191]: I think we can omit the comma here, as the clause following `which` is necessary for which project automation conventions are necessary. A simple test is whether the sentence would make sense without the `which` clause. In this case, the sentence would no longer be accurate as, in general, conventional commits is not necessary for **all** project automation, but only for automation which generates release notes and changelogs.

File: docs/style-guides/git/README.md (hunk 4 above)
Comment (kgryte) [reply docs/style-guides/git/README.md:53]: I will add user stories in a follow-up PR.

**Comment by kgryte** [comment]
This has received 3 approvals. Will merge.

I'll submit a follow-up PR with user stories to help clarify which commit types apply when.

**Pull request merged**